pkg/scheduler/framework/plugins/volumezone/volume_zone.go

```diff
@@ -221,6 +214,126 @@ func (pl *VolumeZone) Filter(ctx context.Context, cs *framework.CycleState, pod
 	}
 
 	node := nodeInfo.Node()
+	status, _ := pl.isSchedulableNode(logger, podPVTopologies, node, pod)
+	return status
+}
+
+func getStateData(cs *framework.CycleState) (*stateData, error) {
+	state, err := cs.Read(preFilterStateKey)
+	if err != nil {
+		return nil, err
+	}
+	s, ok := state.(*stateData)
+	if !ok {
+		return nil, errors.New("unable to convert state into stateData")
+	}
+	return s, nil
+}
+
+func getErrorAsStatus(err error) *framework.Status {
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return framework.NewStatus(framework.UnschedulableAndUnresolvable, err.Error())
+		}
+		return framework.AsStatus(err)
+	}
+	return nil
+}
+
+// EventsToRegister returns the possible events that may make a Pod
+// failed by this plugin schedulable.
+func (pl *VolumeZone) EventsToRegister() []framework.ClusterEventWithHint {
+	return []framework.ClusterEventWithHint{
+		// New storageClass with bind mode `VolumeBindingWaitForFirstConsumer` will make a pod schedulable.
+		// Due to immutable field `storageClass.volumeBindingMode`, storageClass update events are ignored.
+		{Event: framework.ClusterEvent{Resource: framework.StorageClass, ActionType: framework.Add}, QueueingHintFn: pl.isSchedulableAfterStorageClassAdded},
+		// A new node or updating a node's volume zone labels may make a pod schedulable.
+		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.UpdateNodeLabel}, QueueingHintFn: pl.isSchedulableAfterNodeChange},
+		// A new pvc may make a pod schedulable.
+		// Due to fields are immutable except `spec.resources`, pvc update events are ignored.
+		{Event: framework.ClusterEvent{Resource: framework.PersistentVolumeClaim, ActionType: framework.Add}, QueueingHintFn: pl.isSchedulableAfterPersistentVolumeClaimAdded},
```

Comment:
Also, what if StorageClassName is specified later?
Seems such operation is allowed, isn't it? (can you confirm this?)
https://github.com/kubernetes/kubernetes/blob/08c146b143c242741303aa24cbca9900a5eb4c38/pkg/apis/core/validation/validation.go#L2264

If it's allowed, to specify storage class name could resolve this error.
https://github.com/kubernetes/kubernetes/blob/cd742f30fb84eb121ec6589aad337a93081b4fda/pkg/scheduler/framework/plugins/volumezone/volume_zone.go#L125

Reply:
I tried it, but StorageClassName is immutable. I got an error `spec: Forbidden: spec is immutable after creation except resources.requests for bound claims`.
https://github.com/kubernetes/kubernetes/blob/08c146b143c242741303aa24cbca9900a5eb4c38/pkg/apis/core/validation/validation.go#L2279-L2282